src/github_runner_image_builder/store.py

```diff
@@ -0,0 +1,115 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Module for uploading images to shareable storage."""
+
+import logging
+from pathlib import Path
+from typing import cast
+
+import openstack
+import openstack.connection
+import openstack.exceptions
+from openstack.image.v2.image import Image
+
+from github_runner_image_builder.errors import GetImageError, OpenstackError, UploadImageError
+
+logger = logging.getLogger(__name__)
+
+
+def upload_image(cloud_name: str, image_name: str, image_path: Path, keep_revisions: int) -> str:
+    """Upload image to openstack glance.
+
+    Args:
+        cloud_name: The Openstack cloud to use from clouds.yaml.
+        image_name: The image name to upload as.
+        image_path: The path to image to upload.
+        keep_revisions: The number of revisions to keep for an image.
+
+    Raises:
+        UploadImageError: If there was an error uploading the image to Openstack Glance.
+
+    Returns:
+        The created image ID.
+    """
+    with openstack.connect(cloud=cloud_name) as connection:
+        try:
+            image: Image = connection.create_image(
+                name=image_name,
+                filename=str(image_path),
+                allow_duplicates=True,
```

Comment:
Should we allow duplicates ?

Reply:
Yes, the image names can stay the same whilst differing by image ids.